doc/contrib/coding_guide.rst

```diff
@@ -0,0 +1,158 @@
+################
+Coding Guideline
+################
+
+**Contents**
+
+.. contents::
+  :backlinks: none
+  :local:
+
+********************
+C++ Coding Guideline
```

Comment:
are the guidelines around abi compatibility documented somewhere? if not, should we? ideally, we wouldn't want to break abi (for external facing api's) until a major release (or until a minor version release bump). if we go down that path, perhaps, we should enforce abi checks as a part of pull request as well...

Reply:
@sriramch yes, we had some discussion in https://github.com/dmlc/xgboost/pull/4646/files#diff-0ced3dfd44ba156272e6db34a5247462, '

we will add binary compatibility tools after releasing 1.0 (i.e. taking 1.0.0 as the ground truth)

Reply:
@sriramch Let me clarify the release policy. We plan to guarantee ABI compatibility for all C API functions in [c_api.h](https://xgboost.readthedocs.io/en/latest/dev/c__api_8h.html), but not for any other C++ functions or constructs. This is because only the former is exposed to other language bindings, such as Python, R, and JVM packages.